src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/ProjectSetUIPropertyValueActionCore.cs

```diff
@@ -64,21 +65,39 @@ public async Task OnBeforeExecutingBatchAsync(IEnumerable<UnconfiguredProject> t
         {
             foreach (UnconfiguredProject project in targetProjects)
             {
-                if (!_rules.TryGetValue(project.FullPath, out List<IRule> projectRules))
+                if (!_rules.TryGetValue(project.FullPath, out List<IRule> projectRules)
+                    && await project.GetProjectLevelPropertyPagesCatalogAsync() is IPropertyPagesCatalog projectCatalog
+                    && projectCatalog.GetSchema(_pageName) is Rule rule
+                    && rule.GetProperty(_propertyName) is BaseProperty property)
                 {
+                    bool configurationDependent = property.IsConfigurationDependent();
                     projectRules = new List<IRule>();
-
                     IPropertyPageQueryCache propertyPageCache = _queryCacheProvider.CreateCache(project);
-                    if (await propertyPageCache.GetKnownConfigurationsAsync() is IImmutableSet<ProjectConfiguration> knownConfigurations)
+                    if (configurationDependent)
                     {
-                        foreach (ProjectConfiguration knownConfiguration in knownConfigurations.Where(config => config.MatchesDimensions(_dimensions)))
+                        // The property is configuration-dependent; we need to collect the bound rules for
+                        // all matching configurations.
+                        if (await propertyPageCache.GetKnownConfigurationsAsync() is IImmutableSet<ProjectConfiguration> knownConfigurations)
                         {
-                            if (await propertyPageCache.BindToRule(knownConfiguration, _pageName) is IRule boundRule)
+                            foreach (ProjectConfiguration knownConfiguration in knownConfigurations.Where(config => config.MatchesDimensions(_dimensions)))
```

Comment:
Can potentially improve perf here by running the `BindToRule` calls concurrently, rather than serializing them in a `foreach`.

Something like:

```c#
var tasks = knownConfigurations
    .Where(config => config.MatchesDimensions(_dimensions))
    .Select(config => propertyPageCache.BindToRule(config, _pageName))
    .ToList();

await Task.WhenAll(tasks);

projectRules.AddRange(tasks.Select(task => task.GetAwaiter().GetResult());
```

Reply:
The call to `BindToRule` may cause us to call `UnconfiguredProject.LoadConfiguredProjectAsync`, and that method is likely to take a lock. As such, the potential benefits here may not be as great as we could hope.

It may still be a worthwhile optimization, but I want to wait until we have actual data traces before I start with perf optimizations--especially those that introduce concurrency.